### ISSUE TITLE
flyout Menus do not flip sides to prevent clipping

### ISSUE THREAD

**Issue by christianvogt** [patternfly/patternfly-react#5536]
**Describe the issue. What is the expected and unexpected behavior?**

When a flyout Menu is opened near the edge of the viewport, the flyout always opens to the right. This could result in clipping or scrollbars appearing if there isn't enough space to open the menu.

Similar to context menus on a desktop, I expect the menu item to flip to the opposite side or even open upwards to best orient itself to be fully visible. Multiple nested flyouts should continue to orient themselves accordingly.

**Please provide the steps to reproduce. Feel free to link CodeSandbox or another tool.**

Observable today on the components page: https://www.patternfly.org/v4/components/menu
Adjust the browser width small enough that when opening a flyout menu, it is clipped to the right.

**Is this a bug or enhancement? If this issue is a bug, is this issue blocking you or is there a work-around?**
enhancement

**What is your product and what release version are you targeting?**
OpenShift Console

**Comment by tlabaj** [comment]
cc @mcarrano

**Comment by mcarrano** [comment]
I agree that this is a reasonable expectation. What needs to happen to address this @tlabaj ? Can this be handled purely in React or does it require a core change?

**Comment by christianvogt** [comment]
If the flyout menus are positioned using popperjs, you can get the flip behavior nearly for free.

**Comment by jschuler** [comment]
we currently get styles from PF core which positions the flyout. It would be possible to remove the class from react and use popper for positioning, defaulting to the right location.
What about the menu item caret icon, does that need to change depending on where the flyout appears?

**Comment by mcarrano** [comment]
I'm looking at other implementations of fly-out menus and the caret does not change position, so I think always keeping it on the right is fine. @mcoker what are your thoughts on the above? Based on this, would we want to make the updates in core or just use popper for positioning as @jschuler suggests?

**Comment by mcoker** [comment]
@mcarrano it's tough because we're already using popper for positioning when using the `menuAppendTo` prop, so whether we get the styles from core or react, it seems like we'll need to use popper for positioning anyways when using `menuAppendTo`. That said, I would like to see the positions in core for consistency's sake, having the positions available for non-react users. Also part of the design for these menus is a gap between the toggle and menu that comes from core, even when using `menuAppendTo`, so that would need to be added for left/right menus, and if it comes from react, we try to keep as many of the styles as we can in core, and that could lead to inconsistencies if we update the gap in core, but we forget to update it in react. So I'm in favor of adding the styles in core, though I don't feel strongly about it since we'll end up needing to use popper anyways w/ `menuAppendTo`. If we do add this to core, looks like we'll need to add left-top, left-bottom, and right-top and right-bottom? We already have bottom-left (default), bottom-right, top-left, and top-right.

**Comment by mcarrano** [comment]
Thanks, @mcoker I will transfer this to core then and get it on our roadmap.
